src/transforms/aggregate.rs

```diff
@@ -0,0 +1,372 @@
+        agg.record(gauge_a_1.clone());
+        agg.record(gauge_a_2.clone());
+        out.clear();
+        assert_eq!(1, agg.flush_into(&mut out));
+        assert_eq!(1, out.len());
+        assert_eq!(&gauge_a_2, &out[0]);
+
+        let gauge_b_1 = make_metric("gauge_b", metric::MetricKind::Absolute,
+            metric::MetricValue::Gauge { value: 44.0 }, tags.clone());
+        // Two increments with the different series, should get each back as-is
+        agg.record(gauge_a_1.clone());
+        agg.record(gauge_b_1.clone());
+        out.clear();
+        assert_eq!(2, agg.flush_into(&mut out));
+        assert_eq!(2, out.len());
+        // B/c we don't know the order they'll come back
+        for event in out {
+            match event.as_metric().series().name.name.as_str() {
+                "gauge_a" => assert_eq!(gauge_a_1, event),
+                "gauge_b" => assert_eq!(gauge_b_1, event),
+                _ => assert!(false),
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn transform() {
+        let agg = toml::from_str::<AggregateConfig>(
+            r#"
+interval_ms = 10000
+"#,
+        )
+        .unwrap()
+        .build(&GlobalOptions::default())
+        .await
+        .unwrap();
+        let agg = agg.into_task();
+
+        let tags: BTreeMap<String, String> =
+            vec![("tag1".into(), "val1".into())].into_iter().collect();
+        let counter_a_1 = make_metric("counter_a", metric::MetricKind::Incremental,
+            metric::MetricValue::Counter { value: 42.0 }, tags.clone());
+        let counter_a_2 = make_metric("counter_a", metric::MetricKind::Incremental,
+            metric::MetricValue::Counter { value: 43.0 }, tags.clone());
+        let counter_a_summed = make_metric("counter_a", metric::MetricKind::Incremental,
+            metric::MetricValue::Counter { value: 85.0 }, tags.clone());
+        let gauge_a_1 = make_metric("gauge_a", metric::MetricKind::Absolute,
+            metric::MetricValue::Gauge { value: 42.0 }, tags.clone());
+        let gauge_a_2 = make_metric("gauge_a", metric::MetricKind::Absolute,
+            metric::MetricValue::Gauge { value: 43.0 }, tags.clone());
+        let inputs = vec![counter_a_1, counter_a_2, gauge_a_1, gauge_a_2.clone()];
+
+        let in_stream = Box::pin(stream::iter(inputs));
+        let mut out_stream = agg.transform(in_stream);
+
+        tokio::time::pause();
+        // 1s longer than our timeout configured up above
+        tokio::time::advance(Duration::from_secs(11)).await;
+        tokio::time::resume();
+
+        let mut count = 0;
+        while let Some(event) = out_stream.next().await {
+            count += 1;
+            match event.as_metric().series().name.name.as_str() {
+                "counter_a" => assert_eq!(counter_a_summed, event),
+                "gauge_a" => assert_eq!(gauge_a_2, event),
+                _ => assert!(false),
+            };
+        }
```

Comment:
So, this is closer, but not quite there yet.

The crucial bit here is that we actually need to poll the transform's output stream _after_ pausing time, to demonstrate that even though the input values are immediately available, our flush logic is correctly not flushing those items out because the interval hasn't yet elapsed.

We can combine the `poll!` macro from the `futures` crate and `Stream::next` to ensure that the stream does not yet have any items to return:
```rust
assert_eq!(Poll::Pending, poll!(out_stream.next()).await);
```

This code, instead of fully `await`ing for an actual value (or stream end), simply polls the stream one time to get its instantaneous state.  Given that this stream has to be polled/driven to update its internal state, we know that if we poll it once, and we get back `Poll::Pending`, that it is in the correct state.

After that, once we advance the time -- no need to `resume`, actually -- the existing `while` loop confirms that it flushes properly, and we do our remaining checks on the validity of the returned events, and so on.

Reply:
Will look into this. Fwiw this test was cribbed straight from reduce's suite so if something is lacking in it the same is probably true there.

Reply:
The test code flushes due to the "we're shutting things down" state where `input_rx.next() => None` so it explicitly doesn't wait for the interval to elapse and the outputs are effectively there to read immediately. That seems to be the same for `reduce` and its tests which the transform function was ported from. There's no tests there of whether the `poll_period` works afaict they're all relying on the input stream has ended flush.

> The crucial bit here is that we actually need to poll the transform's output stream _after_ pausing time, to demonstrate that even though the input values are immediately available, our flush logic is correctly not flushing those items out because the interval hasn't yet elapsed.

Still looking into this. I think I'll have to manually manage and feed the input stream rather than create one from a vector to get something I can avoid seeing the end/close immediately.

Reply:
OK, pushed up a commit with tests for both the final closing down flush as well as an interval based on. The interval version is pretty involved, but after doing a bunch of tracing/debugging this seems to be what it takes to reliably get things to happen when/as expected.